saleor/discount/tasks.py

```diff
@@ -157,10 +158,18 @@ def fetch_promotion_variants_and_product_ids(promotions: "QuerySet[Promotion]"):
         rule_variants = get_variants_for_predicate(rule.catalogue_predicate)
         variants |= rule_variants
         promotion_id_to_variants[rule.promotion_id] |= rule_variants
-    products = Product.objects.filter(
-        Exists(variants.filter(product_id=OuterRef("id")))
-    )
-    return promotion_id_to_variants, list(products.values_list("id", flat=True))
+
+    product_ids: set[int] = set()
+    variants = variants.order_by("pk")
+    for variant_ids in queryset_in_batches(variants, 2000):
+        variants_batch = ProductVariant.objects.filter(id__in=variant_ids).all()
+        product_ids.update(
+            Product.objects.filter(
+                Exists(variants_batch.filter(product_id=OuterRef("id")))
+            ).values_list("id", flat=True)
+        )
+
+    return promotion_id_to_variants, list(product_ids)
```

Comment:
We still fetch an infinite amount of products, this could cause:
- Payload too big here thus crashing: https://github.com/saleor/saleor/blob/c14ce5cfd8aad483f48b30e370dc0514309df286/saleor/discount/tasks.py#L81
- High memory usage

Also, are we sure Celery beat isn't going to trigger duplicate `update_products_discounted_prices_for_promotion_task` tasks?

Reply:
1. It is not infinite. Max is a number of products in database. Scenario: I want to discount all of my products by 10% is a common case.
2. Yes, payload size can be troublesome.
3. We use `set()` on product_ids and update prices in batches. The `update_products_discounted_prices_for_promotion_task` fires itself, once the batch is processed, so I think we are good here.

Reply:
> It is not infinite. Max is a number of products in database. Scenario: I want to discount all of my products by 10% is a common case.

What I meant is that the value is infinite in code, we don't know how much data this code will have to accept, it can be 10, 10k, 100 billions

> We use set() on product_ids and update prices in batches. The update_products_discounted_prices_for_promotion_task fires itself, once the batch is processed, so I think we are good here.

`set()` doesn't affect how Celery beat works whatsoever, as long as there is no concurrency checks for `handle_promotion_toggle` then it may duplicate tasks